Page Weight 2021 Markdown (#2517)
HTTPArchive/almanac.httparchive.org: src/content/en/2021/page-weight.md

featured_quote: It's critically important to stop optimizing web pages for &dollar;1,200.00 smartphones with ultra-fast internet connections because the vast majority of internet users don't live in that world, and the energy required to power that world is unsustainable.
featured_stat_1: 1,923
featured_stat_label_1: The median number of mobile page bytes
featured_stat_2: 877
featured_stat_label_2: The median number of mobile page image bytes
featured_stat_3: 69
featured_stat_label_3: The median number of mobile page requests
unedited: true
---

> ## Introduction

Reviewer: This is a personal preference, but I kind of like getting a preview of the major findings in the intro.

Another negative side effect of large page weight is increased [latency](https://developer.mozilla.org/en-US/docs/Web/Performance/Understanding_latency) (the time it takes for the request to connect to the server storing the files and begin the process of transporting those files). If bunches of files are requested, no matter the technology, there is a limit on how much can be processed and transferred in any given period. I've audited WordPress sites that request as many as 170 files or more, which ensures terrible page loading performance starting with high latency periods.

There are certainly optimizations that can be done, such as compressing and combining certain file requests, using HTTP/2 or the newer HTTP/3 protocols, and using a modern browser's ability to preconnect to and preload certain files to speed the the whole process process up. The [Performance chapter](https://almanac.httparchive.org/en/2021/performance) covers a wide range of factors that effect page loading performance.

## Assets

> As explained in last years chapter, there's been little change in most assets in terms of technology, but there are some notable exceptions.

Reviewer: Many folks might not have read last year's chapter. You could maybe reword the sentiment to stand alone to something like "Since last year, assets have experienced X major changes." Then a short sentence or bullets summarizing the main points (which are then covered in detail in each of the subsections)

### Static

> Static files reside by themselves and are used as resources to help build out and render web pages like video and audio files, and font files are all examples of static assets. [Images](https://almanac.httparchive.org/en/2021/media), which make a large percentage of the average web page's weight, are also a type of static asset file. So, let's use images for our example.

Reviewer: ```suggestion
Static files reside by themselves and are used as resources to help build out and render web pages. Image, video, audio, and font files are all examples of static assets. [Images](https://almanac.httparchive.org/en/2021/media) make a large percentage of the average web page's weight, so let's use them for our example.
```

I'm confused what the example is for. Or maybe this section? Are you just describing what each asset type is, or were you going to talk about changes from last year to this year? Or both? It's up to you, I'm just saying it might be a little unclear to the reader what the purpose is here.

Consider putting the actual % number for images (as a total of page weight) here as a teaser.

> Image file formats like PNG and JPEG are supported by browser vendors for rendering in web pages. More recent image formats, such as WebP, offer high quality, excellent compression (much better than PNG or JPG), and are supported by all major browsers have gained popularity. Some image types that are very good for the web lack full cross browser support so researching what image format will work with what browser is essential. Moreover, the latest browser features can look for the best choice for the browser requesting that image and defaulting back to a fully supported image if one is not available. The main point about images is to make sure they are optimized for the web, the [Media](https://almanac.httparchive.org/en/2021/media) chapter covers this in much more detail. Failing to properly size and compress images for your site will exact a high price on performance.

Reviewer: ```suggestion
Image file formats like PNG and JPEG are widely supported by all browsers. More recent image formats, such as WebP and AVIF, offer higher quality with smaller file sizes have gained popularity. WebP is supported by most modern browsers, while AVIF is newer and less supported [maybe put caniuse links here]. With the `<picture>` tag, you can use modern image formats while providing JPEG and PNG fallbacks. Make sure your images are optimized for the web-the [Media](https://almanac.httparchive.org/en/2021/media) chapter covers this in much more detail. Failing to properly size and compress images for your site will exact a high price on performance.
```

**Note**: If you need an online service that will optimize your images, there is no better source I've found than Google's [Squoosh](https://squoosh.app/) application, and Jake Archibald's [SVGOMG](https://jakearchibald.github.io/svgomg/) for optimizing SVG's.

> #### A word about the proliferation in the use of JavaScript

Reviewer: Just noting here that I'm going to stop doing "edits" on sections where I get where you're going with the content. For those that I have more "review" stuff, I'll still do some editorial suggestions.

Is this note supposed to be under the "static" subheading? Just double checking because the following section is confusing.

Example: You navigate to a e-commerce site on your nifty smartphone on slow 4G internet connection, but the site is taking forever to load. It even throws a warning that the page is taking too long to load and asks if you want to continue waiting. Your smartphone or computer stalls and may even need to be restarted. Nine times out of ten a blocking JavaScript that is causing your smartphone to run out of processing resources or memory is to blame. The judicious, expert use of JavaScript can create great user experiences. **But remember this**: *JavaScript is executed on the client side. It's using the client computers resources to process and execute the script, and there is a finite amount of resources on every device*. Once again, not everyone is glued to the newest Google Pixel or Apple smartphone. The [JavaScript](https://almanac.httparchive.org/en/2021/javascript) chapter contains a wealth of information about this issue.

> ### Dynamic

Reviewer: I don't think I understand this section

Dynamically generated assets normally are produced by web programming languages like PHP, JavaScript, etc. Like other files, a requests is sent to locate, process, and retrieve the results of the scripts. The payload is returned to the user's browser for rendering. Generally, dynamic scripting in modern web sites execute extremely fast, and their results can also be cached for even faster use in future requests. But poorly scripted dynamic assets can have the opposite affect, increasing page weight. So, for best results, use best practices and and apply optimizations.

> ### Connected

Reviewer: Not sure I understand this one either. Are you saying "connected" is a type of asset? Are you talking more about 3rd party assets?

Page weight can also be effected by external services that are connected to the web page. Some of those services include CDN's, analytics, chat bots, forms, and other data collection and processing methods. I find this to be one of the fastest growing problem areas that result in bloated page weight. Many of these third party services use outdated, poorly written JavaScript and querying techniques that take much longer to execute than they should, and the site owner has little control over how that third party impacts the loading of a page. Suffice it to say that inquiring about how a service will affect your page loading performance is very important. So is testing their impact.

> ### Caching

Reviewer: Caching is a strategy, not a type of asset. I think I'm confused what the broader "Assets" (h2) section is about and how each subheading and subsection relates to it? What's the outline and how does it all tie together?

[Caching](https://almanac.httparchive.org/en/2021/caching), are special services/processes that create copies of previously requested optimized assets and store them on remote server or on the client's browser/disk until needed again. Caching of optimized assets dramatically lowers page weight and page loading time because the asset is immediately available, removing the need to execute and entire request process.

## Analysis

> As we post and parse statistical results, the data is often based on transfer sizes. However, we are employing decompressed sizes in this analysis whenever possible. Please note that many of the results listed below are similar or the same as those queries used in 2020.

Reviewer: I'm unclear what you're saying here. Did something change from last year? Maybe just state directly something like "This analysis shows decompressed sizes, not transfer sizes".

However, that seems odd considering the previous commentary on transmission stuff. Both are important. I'd be curious if there are different things we could infer by looking at both.

### Page weight by the numbers

> Looking at the page weight on both desktop and mobile devices, the deltas are due to the difference in the amount of transferred resources on the mobile device, the difference is generally small between them.

Reviewer: I think you could delete this paragraph.

{{ figure_markup(
  image="bytes-distribution.png",
  caption="Distribution of total bytes per page.",
  description="Bar chart showing the distribution of the total bytes per page. Desktop pages tend to have more bytes throughout the distribution. The 10, 25, 50, 75, and 90th percentiles for mobile pages are: 409, 928, 1,923, 3,749, and 6,890 KB per page.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vThwYNAo2K8eVJVNH_4TuA7IZpG534jXn5vFWphLLuHUo_VNGehDMKoIFWf0GquINRKWNHX66MXUmn9/pubchart?oid=8302830&format=interactive",
  sheets_gid="895329225",
  sql_file="bytes_per_type_2021.sql"
) }}

> The key data here shows we are closing in on 6.9 MB of page weight on mobile and 8.1 MB on desktop at the 90th percentile. The data is following an old and consistent trend: growth in page weight is on the upward trajectory that continues from the previous year.

Reviewer: I would first state the median page weight and compare that to the previous year. You can also compare the 90th percentile to the previous year, but the 90th percentile this year alone does not tell you anything about a trend. If you have a number, maybe state it or add another chart showing the year-over-year change just for mobile devices (for each percentile).

{{ figure_markup(
  image="bytes-distribution-content-type.png",
  caption="Median bytes per page by content type.",
  description="Bar chart showing the median number of bytes per page for images, JavaScript, CSS, and HTML. The median desktop page tends to have more bytes. The median mobile page has 877 KB of images, 470 KB of JS, 66 KB of CSS, and 27 KB of HTML.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vThwYNAo2K8eVJVNH_4TuA7IZpG534jXn5vFWphLLuHUo_VNGehDMKoIFWf0GquINRKWNHX66MXUmn9/pubchart?oid=263743460&format=interactive",
  sheets_gid="895329225",
  sql_file="bytes_per_type_2021.sql"
) }}

> A closer inspection shows that the media and average for each resource, and following the trend from years previous to this one, images remain the largest resource followed by JavaScript.

Reviewer: This sentence is confusing. A possible change:

```suggestion
Images are the largest contributor to page weight, followed by JavaScript.
```

I'd like to see how these numbers compare to total page weight and then see them as percentages of total page weight too. How have these percentages changed year over year?

### Requests

> As previously explained in this chapter, the number of requests and the type of resource requested has a direct influence on total page weight and page loading performance. The difference between current results for this year and last are notable because it actually shows a tiny decrease in the average number of GET requests. Let's hope that trend continues downward.

Reviewer: Is the following chart only for GET requests?

When you mention a trend or trajectory, consider showing a chart with the data.

  sheets_gid="2090334936",
  sql_file="request_type_distribution_2021.sql"
) }}

Images again make up the largest number of requests, though JavaScript is closing in as the gap has narrowed slightly in the last year.

### File formats

{{ figure_markup(
  image="response-distribution-format.png",
  caption="Distribution of image sizes by format.",
  description="Box plot of the distribution of image sizes by format: gif, ico, jpg, png, svg, and webp. JPG sticks out as having the highest distribution with a 90th percentile exceeding 190 KB per image, up from the previous year. WebP has overtaken PNG as second highest at about 93 KB at the 90th percentile. PNG is now third with smaller 90th percentile than WebP, its 75th percentile is higher. gif, ico, and svg all have relatively tiny distributions.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vThwYNAo2K8eVJVNH_4TuA7IZpG534jXn5vFWphLLuHUo_VNGehDMKoIFWf0GquINRKWNHX66MXUmn9/pubchart?oid=1911661099&format=interactive",
  sheets_gid="38001790",
  sql_file="requests_format_distribution.sql"
) }}

> We know that images responsible for a large percentage of web page weight. The above graphic shows the top sources of image weight and the weight distribution. Top 3: JPG, WebP and PNG. So not only is the JPG the most popular image format, it also tends to be the largest by size as well - even larger than a lossless format like the PNG. Likely, the rise in popularity for WebP is due to all major browsers finally supporting the file format. PNG remains popular for use cases such as icons and logos.

Reviewer: This statement is a little misleading:

> So not only is the JPG the most popular image format, it also tends to be the largest by size as well - even larger than a lossless format like the PNG.

In my head, I see this as a good thing because it might mean that people recognize the limited cases where you might want to use a PNG. Because in most circumstances where you use a JPG, the PNG analog would be larger.

### Image bytes

{{ figure_markup(
  image="response-distribution-images.png",
  caption="Distribution of image response sizes per page.",
  description="Bar chart showing the distribution of image bytes per page. Desktop pages tend to load more image bytes per page throughout the distribution. The 10, 25, 50, 75, and 90th percentiles for mobile pages are: 62, 257, 877, 2,324, and 4,992 KB of images per page.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vThwYNAo2K8eVJVNH_4TuA7IZpG534jXn5vFWphLLuHUo_VNGehDMKoIFWf0GquINRKWNHX66MXUmn9/pubchart?oid=1440589973&format=interactive",
  sheets_gid="895329225",
  sql_file="request_type_distribution_2021.sql"
) }}

> Looking at total image bytes shows us that this metric has remained virtually unchanged from the previous year. This is likely due more and more images being served by content distribution networks (CDN), which apply strong optimizations to images as they are uploaded to their servers for use on future requests.

Reviewer: We don't see the previous year here so this does not "show" us that. This chart only shows desktop and mobile at different percentiles for this year.

## Conclusion

> The overall trend moving forward will likely continue to grow for desktop rendered pages and remain mostly unchanged for mobile. Just as they did, along with other browser vendors, for effort to move sites to HTTPS, Google has applied pressure on site owners to put their web pages on a diet, especially for mobile. Google's strategy has been solitified into a program where certain metrics (now called Core Web Vitals) that significantly impact the user experience reward web sites with smaller, faster loading pages by raising their rankings in search results. Conversely, site owners who continue to build heavy pages will be penalized in search results. In a world where small changes in search rankings can have a big impact on the success or failure of a site and their owner's, we should see greater emphasis on creating lighter web pages from design to hosted services that continue to find technological solutions to help keep page assets small and transported in a much faster way.

Reviewer: The first two sentences are really confusing. Also not sure we should suggest a future trend? The rest of the paragraph could use a lot of slimming.

Reviewer 2: Agreed. It should be more like the latest trends suggest. Will fix the conclusion up.

> But will the pressure to lighten the page load, so to speak, cross all web boundaries? What about web titans like Amazon will they follow suit? The answer is unclear. Predicting how the web will be shaped is difficult at best. But there is reason to be optimistic that both kings and subjects in the web world all have good reasons to to make lighter page weights a priority, even if the reasons for doing so differ. A small publisher on the web may be motivated to create ligher pages to compete in search ranking, increase the enjoyment of their users, and doing there part to reduce the amount of energy needed to store, transmit, and render those pages. The Amazon's may want to take advantage of reducing the size of page assets and services to reduce the spend required to serve those pages, and to move into newly emerging markets where users may not be able to buy super fast smartphones or have access to 5G data networks or high speed cable providers. Time will tell.

Reviewer: Might be presuming a lot here. Hard to read also.

Reviewer 2: Yeah, I was really rushed on the closing paragraphs. Will give them a significant update tomorrow.

Reviewer 3: @logicalphase any update on these last few changes?

> What is clear is that the upside of lighter, faster loading web pages provide a strong incentive to commit to take advantage of emerging technologies, design and development techniques, and other solutions to stay ahead of the pack.

Reviewer: This sentence does not make sense to me.

Reviewer 2: Agreed
